design/one-pager-trait-workload-interaction-mechanism.md

```diff
@@ -0,0 +1,112 @@
+accommodate any existing CRD to be imported as `workload` or `trait`. Therefore, we cannot rely
+on any pre-defined CRD fields beyond Kubernetes conventions (i.e. spec or status).  
+
+## Proposal
+The overall idea is for the applicationConfiguration controller to annotate critical information
+in the workload and trait CR it emits. In addition, we will provide a helper library so that
+trait controller developers can locate the resources they need with a simple function call. 
+
+Here is the list of changes that we propose. 
+- Add a `childResources` field in the  WorkloadDefinition. 
+Currently, a workloadDefinition is nothing but a shim of a real workload CRD. We propose to add
+an **optional** field called `childResources` to the schema of the workloadDefinition. We encourage
+workload owners to fill in this field when they register their controllers to the OAM system. 
+This is the way for them to declare the types of the Kubernetes resources their workload
+controller actually generates. For example, the following workload definition allows a workload
+owner to declare that the workload creates a deployment and a service. 
+    ```yaml
+    apiVersion: core.oam.dev/v1alpha2
+    kind: WorkloadDefinition
+    metadata:
+       name: mydb.standard.oam.dev
+    spec:
+      definitionRef:
+         name: mydb.standard.oam.dev
+      childResources:
+         - groupVersion: apps/v1
+           kind: deployment
+         - groupVersion: core/v1
+           kind: service      
+    ```
+- ApplicationConfig controller add annotations containing the workload information to the emitted
+`trait` CR. The application controller actually generates the workload CR so it has all the
+information. For example, the following CR have the GVK and name of the workload the
+`ManualScalerTrait` is applied to.
+    ```yaml
+    apiVersion: core.oam.dev/v1alpha2
+    kind: ManualScalerTrait
+    metadata:
+      annotations:
+        core.oam.dev/workloadGroupVersion: core.oam.dev/v1alpha2
+        core.oam.dev/workloadKind: ContainerizedWorkload
+        core.oam.dev/workloadName: foo
+    ```
+- ApplicationConfig controller add annotations containing workload child resource information to
+the emitted `workload` CR. For each `workload` CR, the applicationConfig controller can fetch the
+corresponding `workloadDefinition` CR following an OAM convention. The conventions says that the
+name of the `workloadDefinition` CR is the name of the `workload` CRD it refers to. For example
+, the name of the `workloadDefinition` CR that refers to a `containerizedworkloads.core.oam.dev`
+CRD is exactly `containerizedworkloads.core.oam.dev` as well. For example, the following workload CR
+have all its child resources GVK in its 
+    ```yaml
+    apiVersion: standard.oam.dev/v1alpha2
+    kind: Mydb
+    metadata:
+      annotations:
+        childResources: "{ apps/v1.deployment, core/v1.service}"
```

Comment:
add `resource name` annotation if a workload can launch multiple k8s resource with same type

Reply:
mentioned it in the proposal but not sure if we need to implement it right away.